lib/stdlib/doc/src/gen_statem.xml

```diff
@@ -1168,6 +1191,46 @@ handle_event(_, _, State, Data) ->
 	      an event inserted this way from any external event.
 	    </p>
 	  </item>
+	  <tag><c>change_callback_module</c></tag>
+	  <item>
+	    <p>
+	      Changes the callback module to
+	      <c><anno>NewModule</anno></c>
+	      which will be used when calling all subsequent
+	      <seealso marker="#state callback">state callbacks</seealso>.
+	    </p>
+	    <p>
+	      The <c>gen_statem</c> engine will find out the
+	      <seealso marker="#type-callback_mode">
+		<em>callback mode</em>
+	      </seealso>
+	      of <c><anno>NewModule</anno></c> by calling
+	      <seealso marker="#Module:callback_mode/0">
+		<c>NewModule:callback_mode/0</c>
+	      </seealso>
+	      before the next
+	      <seealso marker="#state callback">state callback</seealso>.
+	    </p>
+	    <p>
+	      Changing the callback module does not affect the
+	      <em>state transition</em> in any way,
+	      it only changes which module that handles the events.
```

Comment:
From what I understand, that means that you can't change the state or data at the same time. That's probably too limited for my case because I will need to convert the data at the same time (HTTP/1.1 and HTTP/2 have very different state data). I can probably do an intermediate state in order to convert the data before changing the callback module so it's not a blocker. (It can't be done after module change because the new module doesn't have the record definition.)

I am wondering however about the state name itself. In my case the states do not match because when transitioning to HTTP/2 the preface must be sent and therefore if I transition from a `connected` state I can't just go to the corresponding `connected` state for HTTP/2. I also need to do some protocol-specific initialization (not just converting the state), setup timeouts, call callbacks and so on. So it would be ideal if I could do something like this:

``` erlang
statem_callback(..., StateData0=#st{a=A, b=B, c=C...}) ->
    cancel_timeouts_etc(StateData0),
    StateData = new_statem:convert_state(A, B, C...),
    [{next_state, preface, StateData}, {callback_module, new_statem}].
```

And it brings me into the `preface` state of `new_statem` with `StateData` instead of `StateData0`. The bonus I get from this is that I can share the `preface` state between the different entry points to `new_statem` (for HTTP/2, there are 3 different entry points, but as mentioned yesterday one of them would be better done as an intermediate callback module since it needs to juggle with both HTTP/1.1 and HTTP/2 at the same time).

Again nothing impossible, I would just have to have a transition state `switch_protocol` that exists in both protocols and just go `connected` -> (change state data a the same time) -> `switch_protocol` -> (change callback module) -> `switch_protocol` -> `preface`, but if I can do all in one go I'd take it.

Reply:
@essen: Since changing the callback module does not affect the state transition in any way, it is absolutely possible to combine a state transition that changes State and updates Data with changing callback module.  Changing State and updating Data is orthogonal to changing callback module.

```erlang
statem_callback(..., StateData0=#st{a=A, b=B, c=C...}) ->
    cancel_timeouts_etc(StateData0),
    StateData = new_statem:convert_state(A, B, C...),
    {next_state, preface, StateData,
     [{change_callback_module, new_statem}]}.
```
The next callback that will be called is `new_statem:preface(Type, Content, StateData)`.

You can even insert an event in the same state transition to force a call of the new callback module, so it is possible to instead of having an API function `new_statem:convert_state/N` insert `{next_event, internal, convert_state}` to the `preface` state, so that there is a function clause that converts the state instead of an API function.  (If that would be clearer)

Does that solve all the problems in the rest of your post?

Do you have any suggestion for improvement of the documentation wording, to make this important point clearer?

Reply:
> The next callback that will be called is `new_statem:preface(Type, Content, StateData)`.

More accurately the next callback that will be called is `new_statem:callback_mode()`, and then either `new_statem:preface(Type, Content, StateData)`, or `new_statem:handle_event(Type, Content, preface, StateData)` depending on `new_statem`'s callback mode, or even, if state enter calls are used by `new_statem`; `new_statem:preface(enter, statem_callback, StateData)` or `new_statem:handle_event(Type, enter, statem_callback, StateData)` inbetween...

Reply:
Ah I thought perhaps I was wrong about that. No problem at all then.

About inserting an event to update the state that won't work for me because the new statem doesn't have the record definition of the old statem so some conversion needs to be done beforehand (or manually looking into the record, or exposing the record definition, neither of which I'm willing to do). But in theory that would be possible indeed.

The documentation wording is fine, it's just a complex subject and I get easily confused with the different types of events and actions.